src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.cs

```diff
@@ -9,52 +10,270 @@ namespace Microsoft.Maui.Controls.Platform
+
+		Page CurrentPlatformModalPage =>
+			_platformModalPages.Count > 0 ? _platformModalPages[_platformModalPages.Count - 1] : throw new InvalidOperationException("Modal Stack is Empty");
+
+		Page? CurrentPage =>
+			_modalPages.Count > 0 ? _modalPages[_modalPages.Count - 1] : _window.Page;
+
+		// Shell takes care of firing its own Modal life cycle events
+		// With shell you cam remove / add multiple modals at once
+		bool FireLifeCycleEvents => _window?.Page is not Shell;
+
+		partial void InitializePlatform();
 
 		public ModalNavigationManager(Window window)
 		{
 			_window = window;
+			InitializePlatform();
+
+			_window.HandlerChanging += OnWindowHandlerChanging;
+			_window.Destroying += (_, _) =>
+			{
+				_platformModalPages.Clear();
+			};
 		}
 
-		public Task<Page> PopModalAsync()
+		void OnWindowHandlerChanging(object? sender, HandlerChangingEventArgs e)
 		{
-			return PopModalAsync(true);
+			// If the window handler is changing the activity is being recreated
+			// the window activated/resumed event will take care of syncing the platform modals
+			if (e.OldHandler != null)
+			{
+				_platformModalPages.Clear();
+			}
 		}
 
+		public Task<Page?> PopModalAsync()
+		{
+			return PopModalAsync(true);
+		}
 
 		public Task PushModalAsync(Page modal)
 		{
 			return PushModalAsync(modal, true);
 		}
 
-		internal void SettingNewPage()
+		bool syncing = false;
+
+		bool IsModalReady
 		{
-			if (_previousPage != null)
+			get
 			{
-				// if _previousNavModel has been set than _navModel has already been reinitialized
-				if (_previousNavModel != null)
+				return
+#if ANDROID
+					_window.IsActivated &&
+#endif
+					_window?.Page?.Handler != null && _window.Handler != null;
+			}
+		}
+
+		void SyncPlatformModalStack([CallerMemberName] string? callerName = null)
+		{
+			var logger = _window.FindMauiContext(true)?.Services?.CreateLogger<ModalNavigationManager>();
+			SyncPlatformModalStackAsync().FireAndForget(logger, callerName);
+		}
+
+		// This code only processes a single sync action per call.
+		// It recursively calls itself until no more sync actions are left to perform.
+		//
+		// A lot can change during the process of pushing/popping a page
+		// i.e. Users might change the root page during an appearing event.
+		// So, instead of just bull dozing through the whole sync we perform one
+		// sync step then recalculate the state of affairs and then perform another
+		// until no more sync operations are left.
+		// Typically it's always a good idea to re-evaluate after any async operation has completed
+		async Task SyncPlatformModalStackAsync()
+		{
+			if (!IsModalReady || syncing)
+				return;
+
+			bool changed = false;
+
+			try
+			{
+				syncing = true;
+
+				await WindowReadyForModal();
+
+				int popTo = 0;
+
+				for (var i = 0; i < _platformModalPages.Count && i < _modalPages.Count; i++)
+				{
+					if (_platformModalPages[i] != _modalPages[i])
+					{
+						break;
+					}
+
+					popTo = i + 1;
+				}
```

Comment:
Why the need of two variables to go over the arrays? popTo is always going to be index, am I wrong? Seems too complicated. Just create the indexer as popTo outside the for loop so that you can later use it and simplify the logic:
```suggestion
				int popTo = 0;

				for (var popTo = 0; popTo < _platformModalPages.Count && popTo < _modalPages.Count; popTo ++)
				{
					if (_platformModalPages[popTo] != _modalPages[popTo])
						break;
				}
```
I'm not a expert, I might be missing something.